src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py

```diff
@@ -145,24 +146,36 @@ def event_bid_traded(self, *, bid_trade):
                               market_bid.buyer, market_bid.seller)
             assert bid_trade.offer.energy <= market_bid.energy, \
                 f"Traded bid on target market has more energy than the market bid."
-            self.markets.source.accept_bid(
+            source_trade = self.markets.source.accept_bid(
                 updated_bid,
                 energy=bid_trade.offer.energy,
                 seller=self.owner.name,
                 already_tracked=False,
                 trade_rate=updated_bid.price / updated_bid.energy
             )
-            self.delete_forwarded_bids(bid_info)
 
+            self.after_successful_trade_event(source_trade, bid_info)
         # Bid was traded in the source market by someone else
         elif bid_trade.offer.id == bid_info.source_bid.id:
-            if self.owner.name == bid_trade.seller:
-                return
             self.delete_forwarded_bids(bid_info)
         else:
             raise Exception(f"Invalid bid state for IAA {self.owner.name}: "
                             f"traded bid {bid_trade} was not in offered bids tuple {bid_info}")
 
+    def after_successful_trade_event(self, source_trade, bid_info):
+        if source_trade.residual:
+            target_residual_bid = self.bid_trade_residual.pop(bid_info.target_bid.id)
+            assert target_residual_bid.id != source_trade.residual.id, \
+                f"Residual bid from the trade ({source_trade.residual}) is not the same as the" \
+                f" residual bid from event_bid_changed ({target_residual_bid})."
+            assert source_trade.residual.id not in self.forwarded_bids, \
+                f"Residual bid has not been forwarded even though it was transmitted via " \
+                f"event_bid_trade."
+            res_offer_info = BidInfo(source_trade.residual, target_residual_bid)
```

Comment:
I suggest to rename `res_offer_info` into `res_bid_info`.

Reply:
Of course, you are correct. Will do